reporter/collector_reporter.go

```diff
@@ -0,0 +1,559 @@
+		f := profile.Function().AppendEmpty()
+		f.SetName(int64(getStringMapIndex(stringMap, v.name)))
+		f.SetFilename(int64(getStringMapIndex(stringMap, v.fileName)))
+	}
+
+	// When ranging over stringMap, the order will be according to the
+	// hash value of the key. To get the correct order for profile.StringTable,
+	// put the values in stringMap, in the correct array order.
+	stringTable := make([]string, len(stringMap))
+	for v, idx := range stringMap {
+		stringTable[idx] = v
+	}
+
+	for _, v := range stringTable {
+		profile.StringTable().Append(v)
+	}
+
+	// profile.LocationIndices is not optional, and we only write elements into
+	// profile.Location that at least one sample references.
+	for i := int64(0); i < int64(profile.Location().Len()); i++ {
+		profile.LocationIndices().Append(i)
+	}
+
+	profile.SetDuration(endTS - startTS)
+	profile.SetStartTime(startTS)
+
+	return startTS, endTS
+}
+
+// reportProfile sends a profile to the next consumer
+func (r *CollectorReporter) reportProfile(ctx context.Context) error {
+	profiles := pprofile.NewProfiles()
+	rp := profiles.ResourceProfiles().AppendEmpty()
+
+	sp := rp.ScopeProfiles().AppendEmpty()
+
+	pc := sp.Profiles().AppendEmpty()
+	pc.SetProfileID(pprofile.ProfileID(mkProfileID()))
+
+	startTS, endTS := r.setProfile(pc.Profile())
+	pc.SetStartTime(startTS)
+	pc.SetEndTime(endTS)
+
+	if pc.Profile().Sample().Len() == 0 {
+		log.Debugf("Skip sending of profile to collector with no samples")
+		return nil
+	}
+
+	return r.nextConsumer.ConsumeProfiles(ctx, profiles)
+}
+
+// lookupCgroupv2 returns the cgroupv2 ID for pid.
+func (r *CollectorReporter) lookupCgroupv2(pid libpf.PID) (string, error) {
+	id, ok := r.cgroupv2ID.Get(pid)
+	if ok {
+		return id, nil
+	}
+
+	// Slow path
+	f, err := os.Open(fmt.Sprintf("/proc/%d/cgroup", pid))
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+
+	var genericCgroupv2 string
+	scanner := bufio.NewScanner(f)
+	buf := make([]byte, 512)
+	// Providing a predefined buffer overrides the internal buffer that Scanner uses (4096 bytes).
+	// We can do that and also set a maximum allocation size on the following call.
+	// With a maximum of 4096 characters path in the kernel, 8192 should be fine here. We don't
+	// expect lines in /proc/<PID>/cgroup to be longer than that.
+	scanner.Buffer(buf, 8192)
+	var pathParts []string
+	for scanner.Scan() {
+		line := scanner.Text()
+		pathParts = cgroupv2PathPattern.FindStringSubmatch(line)
+		if pathParts == nil {
+			log.Debugf("Could not extract cgroupv2 path from line: %s", line)
+			continue
+		}
+		genericCgroupv2 = pathParts[1]
+		break
+	}
+
+	// Cache the cgroupv2 information.
+	// To avoid busy lookups, also empty cgroupv2 information is cached.
+	r.cgroupv2ID.Add(pid, genericCgroupv2)
+
+	return genericCgroupv2, nil
+}
+
+// addPdataProfileAttributes adds attributes to Profile.attribute_table and returns
+// the indices to these attributes.
+func addPdataProfileAttributes(profile pprofile.Profile,
```

Comment:
With code duplication we miss out on changes like https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/212 in some parts.